cnd/src/network/protocol.rs

```diff
@@ -0,0 +1,180 @@
+use tracing::trace;
+
+// TODO: Do this in a more elegant manner.
+lazy_static::lazy_static! {
+    static ref JSON_INFO_LENGTH: usize = {
+        let swap_id = SwapId::from_str("7f3a105d-ecf2-4cc6-b35c-b4351ac28a34").unwrap();
+
+        let bytes = b"hello world, you are beautiful!!";
+        let secret = Secret::from(*bytes);
+        let secret_hash = secret.hash();
+
+        let info = SecretHashInfo {
+            swap_id,
+            secret_hash,
+        };
+
+        let serialized = serde_json::to_vec(&info).unwrap();
+
+        serialized.len()
+    };
+}
+
+/// The secret hash sharing protocol works in the following way:
+///
+/// - Dialer (Alice) writes the `SecretHashInfo` to the substream.
+/// - Listener (Bob) reads the `SecretHashInfo` from the substream.
+
+/// Represents a prototype for an upgrade to handle the secret hash sharing
+/// protocol.
+#[derive(Debug, Clone)]
+pub struct SecretHashProtocolConfig {
+    info: Option<SecretHashInfo>,
+}
+
+impl SecretHashProtocolConfig {
+    /// Sender sends the secret hash `info` to the peer.
+    pub fn sender(info: SecretHashInfo) -> Self {
+        SecretHashProtocolConfig { info: Some(info) }
+    }
+
+    /// Receiver receives the secret hash from the peer.
+    pub fn receiver() -> Self {
+        SecretHashProtocolConfig { info: None }
+    }
+}
+
+/// Data sent to peer in secret hash protocol.
+#[derive(Clone, Deserialize, Debug, Serialize)]
+pub struct SecretHashInfo {
+    swap_id: SwapId,
+    secret_hash: SecretHash,
+}
+
+impl UpgradeInfo for SecretHashProtocolConfig {
+    type Info = &'static [u8];
+    type InfoIter = iter::Once<Self::Info>;
+
+    fn protocol_info(&self) -> Self::InfoIter {
+        iter::once(b"/comit/swap/secret_hash/1.0.0")
+    }
+}
+
+impl<C> InboundUpgrade<C> for SecretHashProtocolConfig
+where
+    C: AsyncRead + AsyncWrite + Unpin + Send + Sync + 'static,
+{
+    type Output = SecretHashInfo;
+    type Error = anyhow::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn upgrade_inbound(self, mut socket: C, _: Self::Info) -> Self::Future {
+        trace!("Upgrading inbound connection");
+        Box::pin(async move {
+            let message = upgrade::read_one(&mut socket, *JSON_INFO_LENGTH).await?;
+            let mut de = serde_json::Deserializer::from_slice(&message);
+            let info = SecretHashInfo::deserialize(&mut de)?;
+
+            Ok(info)
+        })
+    }
+}
+
+impl<C> OutboundUpgrade<C> for SecretHashProtocolConfig
+where
+    C: AsyncRead + AsyncWrite + Unpin + Send + Sync + 'static,
+{
+    type Output = ();
+    type Error = anyhow::Error;
+    type Future = Pin<Box<dyn Future<Output = Result<Self::Output, Self::Error>> + Send>>;
+
+    fn upgrade_outbound(self, mut socket: C, _: Self::Info) -> Self::Future {
+        trace!("Upgrading outbound connection");
+        Box::pin(async move {
+            let message = serde_json::to_vec(&self.info)?;
```

Comment:
You are serializing an option here. I think we should either unwrap this because it is a bug if it is not set or refactor to have two different structs for the config to make it impossible to call `OutboundUpgrade` if we didn't supply the info.

Reply:
Fixed, not entirely happy with the code though.
```
       if self.info.is_none() {
            return Box::pin(future::err(anyhow::anyhow!(
                "cannot upgrade without secret hash info"
            )));
        }

        let info = self.info.unwrap();
```